Solve SEF canonical issues

andrepereiradasilva opened the pull request:
This Pull Request is related to issues https://github.com/joomla/joomla-cms/issues/9333 and https://github.com/joomla/joomla-cms/issues/9556.

Also to PR https://github.com/joomla/joomla-cms/pull/9559
#### Summary of Changes

Add a better check for adding the canonical html tag.
#### Testing Instructions
1. Apply this patch. With SEF system plugin active
2. Do the following test scenarios:

| Domain in the SEF plugin field | Canonical added by a component | Result |
| --- | --- | --- |
| (empty) | No | The canonical html tag will NOT be added |
| (empty) | Yes | No change (the component added canonical html tag stays the same) |
| Yes | No | The canonical html tag will be added with the new domain typed in the SEF plugin domain field (ex: https://www.example.com/menu-item/). |
| Yes | Yes | The canonical html tag will be the same as added by the component but now with the new domain typed in the SEF plugin domain field. |

andrepereiradasilva committed:
solve sef canocical issues

andrepereiradasilva commented:
@scence @SharkyKZ @infograf768 @ggppdk @chivitli 

Can you test this and check if all issues are resolved?

andrepereiradasilva commented:
@wilsonge if tested ok, IMHO should go to 3.5.1

wilsonge commented on plugins/system/sef/sef.php:
```diff
@@ -31,28 +31,26 @@ class PlgSystemSef extends JPlugin
+	public function onAfterRoute()
 	{
-		$doc = $this->app->getDocument();
+		$doc = JFactory::getDocument();
```
Is there any reason this can't be `$this->app->getDocument()` still?

chivitli replied:
You get a fatal error if you try to access it like that in onAfterRoute event

andrepereiradasilva replied:
yes ... try it and you'll see ... php errors.

it seems if you use the `$this->app->getDocument();` in onAfterRoute event you get problems. Don't know why. Maybe @mbabker knows better why this happens.

BTW that has the whole reason i change it to `onAfterDispatch` earlier. After component rendering you can use already it.

SharkyKZ commented:
If canonical is added only when domain is set, there is no need to revert other changes since in this case SEF plugin should override components anyway. Otherwise components would not use the canonical domain.

In such a case, canonical should simply be the specified domain + current path.

SharkyKZ commented:
By my logic, it would be best to check for existing canonicals before doing anything.

1. If canonical is found and canonical domain is set, replace current domain with canonical domain.
2. If canonical is found and canonical domain is not set, leave it as is.
3. If canonical is not found and canonical domain is set, add a canonical.
4. If canonical is not found and canonical domain is not set, leave it as is or add a possibly incorrect canonical.

andrepereiradasilva commented:
ok so let me try to code something for that.
don't test yet.

andrepereiradasilva committed:
Update sef.php

andrepereiradasilva commented:
ok i changed the code.

How it works now: 

- Back to using onAfterDispatch.

- If a canonical EXISTS in the document when the `onAfterDispatch` is run, the plugin doesn't add a canonical.
This should **solve issues** with components and duplicate canonicals.

- If a canonical DOESN'T EXIST in the document when the `onAfterDispatch` is run and a domain is ADDED in the SEF plugin domain field, it adds the canonical with that domain.

- If a canonical DOESN'T EXIST in the document when the `onAfterDispatch` is run and a domain is NOT ADDED in the SEF plugin domain field, it only adds the canonical if the canonical URI is different than the current uri.

Please check code and test.

Open to any suggestions/improvements regarding this.

andrepereiradasilva committed:
minor change

ggppdk commented:
Nice and simple to allow components to add rel canonical without reverse engineering anything
- will test this

SharkyKZ commented:
It doesn't override component canonicals when canonical domain is set.
Can add this: `foreach ($doc->_links as $canonical => $link)` to get existing canonical link as `$canonical`. Then use it with `JUri::getInstance()` to get only the path. Finally, append the canonical `$domain` to it.

chivitli commented:
For me this is good, I would never override canonical from a custom component. The rest works as described. So, test ok from me.

It still remains an issue though that canonicals generated this way are not real canonicals, so I am not sure if this is useful at all (or that it doesn't have a negative impact).

chivitli commented:
I just realized you would like the plugin to replace the domain part of the custom component - that sounds like a good idea. At present, this is indeed not the case.

andrepereiradasilva commented:
> It doesn't override component canonicals when canonical domain is set.
> Can add this: foreach ($doc->_links as $canonical => $link) to get existing canonical link as $canonical. Then use it with JUri::getInstance() to get only the path. Then prepend the canonical $domain to it.
> EDIT: in this case the original canonical also needs to be unset unset($doc->_links[$canonical]) since we're adding a modified.

> I just realized you would like the plugin to replace the domain part of the custom component - that sounds like a good idea. At present, this is indeed not the case.

I will add this them.

andrepereiradasilva committed:
better sef canonical

andrepereiradasilva commented:
please check now.

chivitli commented:
Test successful, works as described.

andrepereiradasilva commented:
Test instructions updated.

@chivitli if it's ok for you. please mark as "Tested successfully" in "Joomla! Issue Tracker" https://issues.joomla.org/tracker/joomla-cms/9565 (after login the "Test this" button appears).
For more information see https://docs.joomla.org/Testing_Joomla!_patches#Recording_test_results

chivitli commented:
I have tested this item :white_check_mark: successfully on d00b99c5b03c3cb18bfcc24b2632ec38b1f77680<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9565">issues.joomla.org/joomla-cms/9565</a>.</sub>

SharkyKZ commented:
I have tested this item :white_check_mark: successfully on d00b99c5b03c3cb18bfcc24b2632ec38b1f77680<br /><br />Works as expected.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9565">issues.joomla.org/joomla-cms/9565</a>.</sub>

infograf768 commented:
Sorry folks, but I am totally lost...
I am using basic staging.
After this patch, eventhough I have left the Domain field empty in the SEF plugin, I now get a canonical.
```
 <base href="http://localhost:8888/trunkgitnew/contact-component.html" />
  <meta name="description" content="UNE DESCRIPTION" />
  <meta name="generator" content="Joomla! - Open Source Content Management" />
  <title>Contact Component - My very long test site name</title>
  <link href="http://localhost:8888/trunkgitnew/contact-component.html" rel="canonical" />
  <link href="/trunkgitnew/templates/protostar/favicon.ico" rel="shortcut icon" type="image/vnd.microsoft.icon" />
[...]
```
As far as I know, a canonical link is only used to show Search Engine the preferred url when multiple URLs reach the same page 
`A canonical link element is an HTML element that helps webmasters prevent duplicate content issues by specifying the "canonical" or "preferred" version of a web page[1][2] as part of search engine optimization. It is described in RFC 6596, which went live in April 2012.`


Why then create a canonical when only one URL exists?

infograf768 commented:
I do now understand though that we get also a canonical before this patch when loading another page than the home as I have indeed overseen that in my former tests...
I just don't understand the use of it in that case.

chivitli commented:
There are always multiple urls in Joomla. Not only multiple, but an infinite number of them... As mentioned, [Google says](https://support.google.com/webmasters/answer/139066?rd=1#2): _**Mark up the canonical page** and any other variants with a rel="canonical" link element_.

There is no need for this optimization to not show canonical on the canonical page.

infograf768 commented:
Are'nt we going to get a canonical for ANY page loaded this way. I mean when we have another url to get to the same page, we would have a different canonical.

ggppdk commented:
> Are'nt we going to get a canonical for ANY page loaded this way. I mean when we have another url to get to the same page, we would have a different canonical.

Ok the plugin, previously had code:
- not to add rel canonical if canonical URL is same as current URL
- now it WILL happen that it will add 2 rel canonical for 2 pages that are same

But
1. **the effect will be the same as before**, goggle will consider the 2 pages different
because if you do not add rel canonical for 2 urls then the rel canonical is supposed to be the URL itself !

ggppdk commented:
I don't know google having special code to try to detect if 2 pages are the same if rel canonical is missing, furthermore which 2 pages would it compare and why it try to compare them ?

- anyway the code to check if current URL is same as calculated canonical is not directly relevant to this PR, so it can be re-added

SharkyKZ commented:
I do think that in scenario 4 canonical URL should not be added at all since it's going be incorrect most of the time anyways.
>4. If canonical is not found and canonical domain is not set, leave it as is or add a possibly incorrect canonical.

As ggppdk said before, correct canonicals can only be provided by components. Unless route helpers can be called in the plugin dynamically using URL vars, I don't see how this can be fixed at the moment.

andrepereiradasilva commented:
wait i'm the one confused now!
will respond to each one of your comments.

@infograf768 
> After this patch, eventhough I have left the Domain field empty in the SEF plugin, I now get a canonica

It shouldn't. Are you at `http://localhost:8888/trunkgitnew/contact-component.html` URI?

> I do now understand though that we get also a canonical before this patch when loading another page than the home as I have indeed overseen that in my former tests...
> I just don't understand the use of it in that case.

> Are'nt we going to get a canonical for ANY page loaded this way. I mean when we have another url to get to the same page, we would have a different canonical.

No, we will only get canonical on the pages in which the canoncial is different from the current url. Example:
- `https://domain.tlg/menu-item/?aaaaa`will get canonical
- `https://domain.tlg/menu-item/`will get canonical
See https://github.com/andrepereiradasilva/joomla-cms/blob/patch-6/plugins/system/sef/sef.php#L79-L82

@chivitli
> There are always multiple urls in Joomla. Not only multiple, but an infinite number of them... As mentioned, Google says: Mark up the canonical page and any other variants with a rel="canonical" link element.
> There is no need for this optimization to not show canonical on the canonical page.

Actually that's true.
I can remove the check in https://github.com/andrepereiradasilva/joomla-cms/blob/patch-6/plugins/system/sef/sef.php#L79-L82 if you guys think it's better.

@ggppdk
> Ok the plugin, previously had code: not to add rel canonical if canonical URL is same as current URL
> Now it WILL happen that it will add 2 (different) rel canonical for 2 pages that are same

Yes, before this PR. This is fixed by this PR.

> But the effect will be the same as before, goggle will consider the 2 pages different because if you do not add rel canonical for 2 urls then the rel canonical is supposed to be the URL itself !

Don't understand.

> I don't know google having special code to try to detect if 2 pages are the same if rel canonical is missing, furthermore which 2 pages would it compare and why it try to compare them ?
> anyway the code to check if current URL is same as calculated canonical is not directly relevant to this PR, so it can be re-added

That code is already in this PR. See https://github.com/andrepereiradasilva/joomla-cms/blob/patch-6/plugins/system/sef/sef.php#L79-L82.

@SharkyKZ 
> I do think that in scenario 4 canonical URL should not be added at all since it's going be incorrect most of the time anyways. Better no canonical than wrong canonical.

What scenario 4? Don't understand. Do you mean `If canonical is not found and canonical domain is not set, leave it as is or add a possibly incorrect canonical.`
If so, that code is already in this PR. See https://github.com/andrepereiradasilva/joomla-cms/blob/patch-6/plugins/system/sef/sef.php#L79-L82.

> As ggppdk said before, correct canonicals can only be provided by components. Unless route helpers can be used in the plugin dynamically using URL vars, I don't see how this can be fixed at the moment.

This PR does not overwrite the component added canonical (only if you set the domain in the domain field, and even then only changes the domain).

SharkyKZ commented:
Yes, currently it adds incorrect canonical. Perhaps should just leave it as is (i.e. don't add canonical at all). Wrong canonical URL beats the purpose of canonical tag, to be honest.

To clarify, this is because `JRoute::_('index.php?' . http_build_query($this->app->getRouter()->getVars()), false);` builds non-canonical URL. So the "canonical" URL isn't really canonical, which makes it pointless and maybe even harmful (in sense of search engines).

andrepereiradasilva commented:
> To clarify, this is because JRoute::_('index.php?' . http_build_query($this->app->getRouter()->getVars()), false); builds non-canonical URL. So the "canonical" URL isn't really canonical, which makes it pointless and maybe even harmful (in sense of search engines).

To understand better, can you give examples where the canonical is wrongly built with that code?

Is this case, at least is good
Example:

- https://domain.tlg/menu-item/?aaaaa will get canonical
- https://domain.tlg/menu-item/ will not get canonical

BTW an easy way to check if the canonical is added in thsi PR is add in this line https://github.com/joomla/joomla-cms/pull/9565/files#diff-78cd3849e61af8fc9737ca40ef91ba05R84
```php
echo 'Current: ' . rawurldecode($uri->toString()) . ' | Canonical: ' . htmlspecialchars($canonical);
```

Then you can navigate trough the site, when that message appears, a canonical exists.

SharkyKZ commented:
But the canonical added on https://domain.tlg/menu-item/?aaaaa is https://domain.tlg/menu-item/?aaaaa, which is clearly incorrect.That is the problem. It should be https://domain.tlg/menu-item, but we cannot simply set this in the system-wide SEF plugin. It is possible by checking for component and view vars and loading respective route helpers to build correct URLs but I don't think this change would be approved by the team.

andrepereiradasilva commented:
> But the canonical added on https://domain.tlg/menu-item/?aaaaa is https://domain.tlg/menu-item/?aaaaa, which is clearly incorrect.That is the problem. 

Ok, right i see what you mean now. 
Actually the canonical will be `https://domain.tlg/menu-item/?aaaaa=`
I notice a lot of cases where that happens.
So i agree, it should be disabled them.

> It should be https://domain.tlg/menu-item, but we cannot simply set this in the system-wide SEF plugin. It could be possible by checking for component and view vars and loading respective route helpers to build correct URLs but I don't think this change would be approved by the team.

That seems to be a router issue, not the canonical generation. So this will not come in this PR.

andrepereiradasilva committed:
add only canonical if the sef plugin domain field is entered

andrepereiradasilva committed:
add hint to field

SharkyKZ commented:
Yeah, I understand that. This is why canonical should not be added in the plugin like this. But this needs to be discussed with the team, I think.

andrepereiradasilva commented:
ok. updated the code, now it ONLY adds (or changes in the case a component already added) the canonical if the SEF plugin domain field has some value.
Also added a placeholder (hint) to the domain field.

Test instructions updated in the first post, please test now.

andrepereiradasilva committed:
just better comments

SharkyKZ commented:
Since canonical URL is now used only to canonize domains, it can be left as current path with canonical domain prepended. Router URL is incorrect anyways, so no point in leaving it there.

andrepereiradasilva commented:
right, you're right will change it

andrepereiradasilva committed:
use juri instead of router to avoid wrong canonicals

andrepereiradasilva commented:
done please retest and mark the test result again.

SharkyKZ commented:
I have tested this item :white_check_mark: successfully on 86f63a650f43c26c965f577e6025e55e53c3e3c1<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9565">issues.joomla.org/joomla-cms/9565</a>.</sub>

chivitli commented:
I have tested this item :white_check_mark: successfully on 86f63a650f43c26c965f577e6025e55e53c3e3c1<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9565">issues.joomla.org/joomla-cms/9565</a>.</sub>

andrepereiradasilva commented:
@infograf768 ok for you now too?

wilsonge commented on plugins/system/sef/sef.xml:
```diff
@@ -22,6 +22,7 @@
 				<field name="domain" type="url"
 					description="PLG_SEF_DOMAIN_DESCRIPTION"
 					label="PLG_SEF_DOMAIN_LABEL"
+					hint="https://www.domain.tld"
```
If this is an example link can you make sure it uses example.com please :) thanks!

andrepereiradasilva replied:
done

andrepereiradasilva committed:
make george happy

andrepereiradasilva commented:
Just a little change at @wilsonge request, no need to retest.

infograf768 commented:
I have tested this item :white_check_mark: successfully on 99b82a7807b6792db93548c8d26f05d083211051<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9565">issues.joomla.org/joomla-cms/9565</a>.</sub>

infograf768 commented:
It is now behaving as should. Thanks!

RTC

@wilsonge 

You can merge into 3.5.1<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9565">issues.joomla.org/joomla-cms/9565</a>.</sub>

wilsonge commented:
Merged - thanks :)

andrepereiradasilva commented:
@SharkyKZ @chivitli @infograf768 @ggppdk @wilsonge thanks for tests and/or contributions.
